internal/keyrotation/rotate.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package keyrotation implements the API handlers for running key rotation jobs.
+package keyrotation
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/logging"
+	"go.opencensus.io/trace"
+)
+
+func (s *Server) handleRotateKeys(ctx context.Context) http.HandlerFunc {
+	logger := logging.FromContext(ctx).Named("keyrotation.HandleRotate")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		_, span := trace.StartSpan(r.Context(), "(*keyrotation.handler).ServeHTTP")
+		defer span.End()
+
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		if err := s.doRotate(ctx); err != nil {
+			logger.Errorw("failed to rotate", "error", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		logger.Info("key rotation complete.")
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func (s *Server) doRotate(ctx context.Context) error {
+	metrics := s.env.MetricsExporter(ctx)
+
+	_, allowed, err := s.revisionDB.GetAllowedRevisionKeys(ctx)
+	if err != nil {
+		return fmt.Errorf("rotate-keys unable to read revision keys: %w", err)
+	}
+
+	// First allowed is newest due to sql orderby.
+	if len(allowed) == 0 || time.Since(allowed[0].CreatedAt) >= s.config.NewKeyPeriod {
+		if _, err := s.revisionDB.CreateRevisionKey(ctx); err != nil {
+			return fmt.Errorf("failed to create revision key: %w", err)
+		}
+		metrics.WriteInt("revision-keys-created", true, 1)
+	}
+
+	if len(allowed) < 2 {
```

Comment:
shouldn't be 2 - needs to be time based

A key needs to be allowed for a specific min duration (config, but I think 15d is default)
Once a key is no longer the primary and was created < 15d ago, it can be destroyed.

Reply:
It only deletes if older than s.config.DeleteOldKeyPeriod... It also always keeps at least two. We can drop that if we don't need any kind of minimum?

Reply:
ah, the minimum number of keys needed is 1.

We just need to not destroy the key material if the revision token might be accepted (minted less than 15d ago). Once a token is that old, we don't need to be able to unlock it any longer (data is useless).

Reply:
I'm saying suppose all keys are older than two weeks. Should we destroy all of them or is there any use in keeping around the one active one? That should never happen, but trying to make the code safe.